state/virtualhostkeys.go

```diff
@@ -0,0 +1,109 @@
+package state
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"encoding/pem"
+
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/virtualhostkeys"
+	"github.com/juju/mgo/v3"
+	"github.com/juju/mgo/v3/txn"
+)
+
+// VirtualHostKey represents the state of a virtual host key.
+type VirtualHostKey struct {
+	st  *State
+	doc virtualHostKeyDoc
+}
+
+type virtualHostKeyDoc struct {
+	DocId   string `bson:"_id"`
+	Id      string `bson:"id"`
+	HostKey []byte `bson:"hostkey"`
+}
+
+// Id returns the virtual host key's ID.
+func (s *VirtualHostKey) Id() string {
+	return s.doc.Id
+}
+
+// HostKey returns the virtual host key.
+func (s *VirtualHostKey) HostKey() []byte {
+	return s.doc.HostKey
+}
+
+func newVirtualHostKeyDoc(hostKeyID string) (virtualHostKeyDoc, error) {
+	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return virtualHostKeyDoc{}, errors.Annotate(err, "failed to generate ED25519 key")
+	}
+
+	pemKey, err := gossh.MarshalPrivateKey(privateKey, "")
+	if err != nil {
+		return virtualHostKeyDoc{}, errors.Annotate(err, "failed to marshal private key")
+	}
+	return virtualHostKeyDoc{
+		// Unsure on docID vs Id.
```

Comment:
Currently `virtualHostKeysC` is a model collection, so all the doc id's are automatically namespaced. You just need to convert them back to the natural key with`st.localID(doc.DocID)`.

Reply:
Ah thanks, that clears it up